### PR TITLE
Fixed #33264 -- Made test runner return non-zero error code for unexpected successes.

### PR DESCRIPTION
**Pull request by bmispelon**

### CODE REVIEW

File: docs/topics/testing/overview.txt [hunk 1]
```diff
@@ -331,7 +331,8 @@ but it's pretty intuitive. You can consult the documentation of Python's
 :mod:`unittest` library for details.
 
 Note that the return code for the test-runner script is 1 for any number of
-failed and erroneous tests. If all the tests pass, the return code is 0. This
+failed tests (whether the failure was caused by an error, a failed assertion or
+an unexpected success). If all the tests pass, the return code is 0. This
```
Comment (felixxm) [inline docs/topics/testing/overview.txt:335]: Comma is missing before 'or'/'and' (see https://en.wikipedia.org/wiki/Serial_comma).
```suggestion
failed tests (whether the failure was caused by an error, a failed assertion,
or an unexpected success). If all the tests pass, the return code is 0. This
```

Comment (felixxm) [reply docs/topics/testing/overview.txt:335]: I would add `versionchanged` annotation, e.g.:
```
.. versionchanged: 4.1

    In older versions, Django test runner does not take unexpected successes,
    from tests marked with the `expectedFailure()` decorator, into account in
    the return code.

.. _speeding-up-tests-auth-hashers:
```

File: docs/releases/4.1.txt [hunk 2]
```diff
@@ -249,7 +249,10 @@ Upstream support for MariaDB 10.2 ends in May 2022. Django 4.1 supports MariaDB
 Miscellaneous
 -------------
 
-* ...
+* The :djadmin:`test` command returns a non-zero error code for unexpected
+  successes (passing tests marked with :py:func:`unittest.expectedFailure`).
+  Prior to this change, the output would be marked as "FAILED" but the error
+  code was zero, indicating a success.
```
Comment (felixxm) [inline docs/releases/4.1.txt:255]: Maybe:
```suggestion
* Django test runner now returns returns a non-zero error code for unexpected
  successes from tests marked with :py:func:`unittest.expectedFailure`
  decorator.
```

File: tests/test_runner/test_discover_runner.py [hunk 3]
```diff
@@ -644,6 +644,24 @@ def test_log_logger(self):
...
+    def test_suite_result_with_failure(self):
+        cases = [
+            (1, 'FailureTestCase'),
+            (1, 'ErrorTestCase'),
+            (0, 'ExpectedFailureTestCase'),
+            (1, 'UnexpectedSuccessTestCase'),
+        ]
+        for expected_failures, testcase in cases:
+            with self.subTest(testcase=testcase):
+                runner = DiscoverRunner(verbosity=0)
```
Comment (felixxm) [inline tests/test_runner/test_discover_runner.py:656]: We don't need to re-instantiate a runner
```suggestion
        cases = [
            (1, 'FailureTestCase'),
            (1, 'ErrorTestCase'),
            (0, 'ExpectedFailureTestCase'),
            (1, 'UnexpectedSuccessTestCase'),
        ]
        runner = DiscoverRunner(verbosity=0)
        for expected_failures, testcase in cases:
            with self.subTest(testcase=testcase):
```

File: tests/test_runner/test_discover_runner.py [hunk 4]
```diff
@@ -644,6 +644,24 @@ def test_log_logger(self):
...
+                with captured_stdout(), captured_stderr():
+                    suite = runner.build_suite([
+                        f'test_runner_apps.failures.tests_failures.{testcase}',
+                    ])
+                    result = runner.run_suite(suite)
```
Comment (felixxm) [inline tests/test_runner/test_discover_runner.py:661]: `captured_stdout()` is unnecessary:
```suggestion
                suite = runner.build_suite([
                    f'test_runner_apps.failures.tests_failures.{testcase}',
                ])
                with  captured_stderr():
                    result = runner.run_suite(suite)
```

**Review by felixxm — COMMENTED** [review]
@bmispelon Thanks :+1:

**Review by blockchainGuru1018 — APPROVED** [review]

**Comment by felixxm** [comment]
@blockchainGuru1018 Please stop approving everything.

**Comment by blockchainGuru1018** [comment]
> @blockchainGuru1018 Please stop approving everything.

I approved once review the PRs @felixxm

**Comment by felixxm** [comment]
@blockchainGuru1018 I'm impressed that you review in detail 18 PRs in 20 minutes :clap:

**Comment by blockchainGuru1018** [comment]
> @blockchainGuru1018 I'm impressed that you review in detail 18 PRs in 20 minutes 👏

To be honest, it was taken around 3 hours

File: docs/releases/4.1.txt (hunk 2 above)
Comment (bmispelon) [reply docs/releases/4.1.txt]: I added a "The" at the beginning of the sentence, removed the "returns" that was there twice, and removed the word "decorator" (I figured people either know what `expectedFailure` is or they can click the link and learn that it's normally used as a decorator).

**Comment by bmispelon** [comment]
@felixxm Thanks for the review ✨ 

I committed 3/4 of your suggested edits directly but I tweaked the 4th one (see comment for detail).

**Review by blockchainGuru1018 — APPROVED** [review]

**Comment by felixxm** [comment]
@bmispelon Thanks for updates :+1: 

I squashed commits and added `versionchanged` annotation.

**Comment by blockchainGuru1018** [comment]
> @bmispelon Thanks for updates 👍
> 
> I squashed commits and added `versionchanged` annotation.

great!

**Pull request merged**